diff --git a/Library/Formula/libmagic.rb b/Library/Formula/libmagic.rb
index 2490664..1b3a83d 100644
--- a/Library/Formula/libmagic.rb
+++ b/Library/Formula/libmagic.rb
@@ -16,7 +16,8 @@ class Libmagic < Formula
     p = []
     p << DATA if MacOS.version < :lion
     # Fixes generaly incorrect code, plus issues with certain
-    # versions of cpp. Not fixed in upstream yet.
+    # versions of cpp. Not fixed in upstream yet. Reported as
+    # http://bugs.gw.com/view.php?id=270
     p << 'https://gist.github.com/kwilczynski/6108866/raw/15d1f0a2dba5da23e820b503e8551346a4d235eb/file-5.14.diff'
   end
